### ISSUE TITLE
Add shortcut for "Show only this branch"

### ISSUE THREAD

**Issue by adamshakhabov** [gitextensions/gitextensions#8441]
## Feature description
I use "Show only this branch" command a lot of and currently, to achieve it the below steps should be executed:
1) Hover on the branch name;
2) Right-click;
3) Hover on "Show only this branch" label;
4) Left-Click.

![image](https://user-images.githubusercontent.com/5969304/92585730-04f2d880-f29e-11ea-9032-69b4e686327f.png)

Would be good 
- to have some shortcut, for example, **Shift+Left-Click on the branch name** which displays commit history only of selected branch.
-  or to have a new checkbox setting like **Show only this branch by left mouse click** whose checked value by default force to use "Show only this branch" when a user clicks on a branch name.

## Environment

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.14393.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

**Comment by RussKie** [comment]
> * or to have a new checkbox setting like **Show this branch only by left mouse click** whose checked value by default force to use "Show this branch only" when a user clicks on a branch name.

I don't think this is a good idea, users will turn it on and forget, or turn it on my mistake, and will find the behavior frustrating. We had similar issues with "show first parents" option in the past.

> * to have some shortcut, for example, **Shift+Left-Click on the branch name** which displays commit history only of selected branch.

This option sounds better, though I'm unsure about <kbd>SHIFT</kbd>, it is used as a conjunctive key, i.e. "add to a selection".
May be <kbd>ALT</kbd> + <kbd>Left MB</kbd>? 🤔

**Comment by adamshakhabov** [comment]
Yep, `ALT` + `Left MB` is also good.

**Comment by adamshakhabov** [comment]
The issue can be assigned to me.

**Comment by gerhardol** [comment]
Note that #8452 allocates <kbd>ALT</kbd>+<kbd>Left MB</kbd> to _grid_ events for _Highlight selected branch_ (<kbd>CTRL</kbd>+<kbd>Shift</kbd>+<kbd>B</kbd>), this was about _Show current branch only_ (<kbd>CTRL</kbd>+<kbd>Shift</kbd>+<kbd>U</kbd>).

The handling from left panel should be consistent.
I find Highlight more interesting, but that does not make sense in the side panel if the branch is hidden. Highlight should be added to the context menu after #8414.

**Comment by RussKie** [comment]
Thank you @gerhardol, for pointing it out.
I now realise that having <kbd>ALT</kbd>+<kbd>Left MB</kbd> to highlight a selected branch (invoked in the revision grid), _AND_ to only show a given branch (in the left panel) would be very confusing.

So may be we do the following in the left panel?
- <kbd>ALT</kbd>+<kbd>Left MB</kbd> to highlight a given branch 
- <kbd>CTRL</kbd>+<kbd>Left MB</kbd> to show a given branch

Thoughts?

**Comment by gerhardol** [comment]
> So may be we do the following in the left panel?

OK for me, but we are allocating these shortcuts to these actions "forever" by this so some more thought would be good before a merge...

**Comment by adamshakhabov** [comment]
I also agree.

**Comment by mstv** [comment]
> This option sounds better, though I'm unsure about <kbd>Shift</kbd>, it is used as a conjunctive key, i.e. "add to a selection".

Yes. This applies to <kbd>CTRL</kbd>, too, which can already be used to select multiple branches.

I find "highlight branch" less intrusive than "Show this branch only". So the latter could require an additional modifier key:
- <kbd>ALT</kbd>+<kbd>Left MB</kbd> to highlight a given branch 
- <kbd>CTRL</kbd>+<kbd>ALT</kbd>+<kbd>Left MB</kbd> to show only a given branch
- <kbd>Left MB</kbd> to select a given branch - cancelling "Show this branch only"
- <kbd>CTRL</kbd>+<kbd>Left MB</kbd> to toggle the selection of a given branch
- <kbd>F5</kbd> reset "Show this branch only" as side-effect of refreshing

**Comment by RussKie** [comment]
> I find "highlight branch" less intrusive than "Show this branch only". So the latter could require an additional modifier key:
> 
> * ALT+Left MB to highlight a given branch
> * CTRL+ALT+Left MB to show only a given branch
> * Left MB to select a given branch - cancelling "Show this branch only"
> * CTRL+Left MB to toggle the selection of a given branch
> * SHIFT+CTRL+A to reset "Show this branch only" (and highlight) as side-effect of showing all branches (including refresh), or
>   Esc as mentioned in #8453 (hidden function)

That's a mouth full :D But I think I can buy into this.
Except for the <kbd>ESC</kbd>, see https://github.com/gitextensions/gitextensions/issues/8453#issuecomment-698897240

**Comment by RussKie** [comment]
I've been thinking over night, and there are two thoughts.
1. The left branch provides us with an opportunity to make filtering significantly better and more usable, replacing buggy and difficult to use the filter in the toolstrip.
2. I tried to remember where else I saw add to/remove from selection. Only graphics apps, such as Paint.NET and Photoshop came to mind.
In those <kbd>CTRL</kbd> is used to "add to selection" and <kbd>ALT</kbd> is used to "subtract from selection" ([#1](https://www.getpaint.net/doc/latest/SelectionTools.html), [#2](https://helpx.adobe.com/photoshop/how-to/add-to-subtract-from-selections.html#:~:text=To%20add%20to%20a%20selection%2C%20click%20the%20Add,keyboard%20as%20you%20make%20another%20selection.%20See%20More))

With this in mind, I think we should reserve these for the similar actions, i.e. when we filter branches (e.g. show current branch), with <kbd>CTRL</kbd>+<kbd>Left MB</kbd> we can add another branch to the filter, and use <kbd>ALT</kbd>+<kbd>Left MB</kbd> to remove a branch from the filter.
For the keyboard only users (to make it more accessible) it could be <kbd>CTRL</kbd>+<kbd>F</kbd> to add, and <kbd>ALT</kbd>+<kbd>F</kbd> to remove from the filter.

As an alternative, we could use <kbd>Right MB</kbd> for "remove" action, e.g.:
- add to the filter: <kbd>CTRL</kbd>+<kbd>Left MB</kbd> / <kbd>CTRL</kbd>+<kbd>+</kbd>
- remove from the filter: <kbd>CTRL</kbd>+<kbd>Right MB</kbd> / <kbd>CTRL</kbd>+<kbd>-</kbd>

If we show all branches, <kbd>CTRL</kbd>+<kbd>Left MB</kbd> will toggle "Show this branch only".
If the filter applies to more than one branch, perhaps <kbd>CTRL</kbd>+<kbd>ALT</kbd>+<kbd>Left MB</kbd> will reset the filter to "Show this branch only". Otherwise it acts the same as <kbd>CTRL</kbd>+<kbd>Left MB</kbd>.

To reset the filter we can add another menu to the context menu, or as a left panel toolstrip button, that appears when the filter is on.


Once we decide which way we want to go, we can decide on the highlighting accelerators. E.g. we could use <kbd>ALT</kbd>+<kbd>Left MB</kbd> or <kbd>ALT</kbd>+<kbd>Right MB</kbd>.


Thoughts?

**Comment by mstv** [comment]
We don't draw pictures, rather compare with the Explorer and ListView: <kbd>CTRL</kbd>+<kbd>Left MB</kbd> and <kbd>CTRL</kbd>+<kbd>Space</kbd> _toggle_ the selection.
And that's how it has already been implemented in GE for mouse clicks.
I stick to my proposal (excluding <kbd>Esc</kbd>).
(<kbd>Alt</kbd>+<kbd>Letter</kbd> hotkeys should be reserved for mnemonics.)

**Comment by RussKie** [comment]
I'm going to do some more thinking. I'm not 100% happy with the current proposal.
